src/GafferImage/Premultiply.cpp

```diff
@@ -0,0 +1,122 @@
+//  PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+//  PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
+//  LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+//  NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+//  SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+//////////////////////////////////////////////////////////////////////////
+
+#include "Gaffer/Context.h"
+
+#include "GafferImage/Premultiply.h"
+
+using namespace IECore;
+using namespace Gaffer;
+
+namespace GafferImage
+{
+
+IE_CORE_DEFINERUNTIMETYPED( Premultiply );
+
+size_t Premultiply::g_firstPlugIndex = 0;
+
+Premultiply::Premultiply( const std::string &name )
+	:	ChannelDataProcessor( name )
+{
+	storeIndexOfNextChild( g_firstPlugIndex );
+	addChild( new StringPlug( "byChannel", Gaffer::Plug::In, "A" ) );
+}
+
+Premultiply::~Premultiply()
+{
+}
+
+Gaffer::StringPlug *Premultiply::byChannelPlug()
+{
+	return getChild<StringPlug>( g_firstPlugIndex );
+}
+
+const Gaffer::StringPlug *Premultiply::byChannelPlug() const
+{
+	return getChild<StringPlug>( g_firstPlugIndex );
+}
+
+void Premultiply::affects( const Gaffer::Plug *input, AffectedPlugsContainer &outputs ) const
+{
+	ChannelDataProcessor::affects( input, outputs );
+
+	if( input == inPlug()->channelDataPlug() ||
+	    input == byChannelPlug() )
+	{
+		outputs.push_back( outPlug()->channelDataPlug() );
+	}
+
+}
+
+void Premultiply::hashChannelData( const GafferImage::ImagePlug *output, const Gaffer::Context *context, IECore::MurmurHash &h ) const
+{
+	ChannelDataProcessor::hashChannelData( output, context, h );
+
+	inPlug()->channelDataPlug()->hash( h );
+	byChannelPlug()->hash( h );
+}
+
+void Premultiply::processChannelData( const Gaffer::Context *context, const ImagePlug *parent, const std::string &channel, FloatVectorDataPtr outData ) const
+{
+	if ( channel == byChannelPlug()->getValue() ) return;
+
+	ConstStringVectorDataPtr inChannelsData = inPlug()->channelNamesPlug()->getValue();
+	const std::vector<std::string> &inChannels = inChannelsData->readable();
+
+	if ( std::find( inChannels.begin(), inChannels.end(), byChannelPlug()->getValue() ) == inChannels.end() )
+	{
+		std::ostringstream channelError;
+		channelError << "Channel '" << byChannelPlug()->getValue() << "' does not exist";
+		throw( IECore::Exception( channelError.str() ) );
+	}
+
+	ContextPtr tmpContext = new Context( *context, Context::Borrowed );
+	tmpContext->set( ImagePlug::channelNameContextName, byChannelPlug()->getValue() );
+	Context::Scope scopedContext( tmpContext.get() );
+
+	const std::vector<float> &a = inPlug()->channelDataPlug()->getValue()->readable();
+	std::vector<float> &out = outData->writable();
+
+	std::vector<float>::const_iterator aIt = a.begin();
+	for ( std::vector<float>::iterator outIt = out.begin(); (outIt != out.end()) && (aIt != a.end()); outIt++, aIt++ )
```

Comment:
For loops like this, I've measured definite overhead in repeatedly fetching `end()` rather than fetching it into a variable at the start. Also you can assume that the length of the alpha data will match the length of the channel data, so just use the one test. I'm not sure it'll have any performance effect in this case, but I'd also use `++it` rather than `it++` - it's worth getting into the habit as some of Gaffer's own iterators are noticeably quicker with the prefix form, and there's no downside.